File: frontend/src/types/inventory.ts
/** Mirrors backend/app/schemas/inventory.py. */

export type InventoryItemType = 'product' | 'raw_material'
// The manual adjustment form only ever submits these four (see
// StockAdjustPayload below) -- receiving, fulfilment, production and
// supplier returns each write their own movement_type through their own
// service, never through this form.
export type MovementType = 'receipt' | 'issue' | 'adjustment' | 'return'
// The full set a StockMovement row can actually carry (backend
// app/models/inventory.py's movement_type enum) -- broader than
// MovementType above, since most of these come from other flows, not
// the manual adjustment form. 'reserve'/'release' are the reservation
// ledger entries written by reserve_stock/release_reservation.
export type AnyMovementType =
  | 'receipt'
  | 'issue'
  | 'adjustment'
  | 'production_in'
  | 'production_out'
  | 'return'
  | 'return_to_supplier'
  | 'reserve'
  | 'release'

export interface StockLevel {
  item_type: InventoryItemType
  item_id: number
  quantity_on_hand: number
  quantity_reserved: number
  quantity_available: number
  /** Set only when this StockLevel was returned by an action that just
   * created a movement -- the ledger entry that adjustment is traceable
   * to. null from a plain GET. */
  movement_id: number | null
}

export interface StockAdjustPayload {
  item_type: InventoryItemType
  item_id: number
  quantity: number
  movement_type: MovementType
  // Mandatory for every manual adjustment.
  notes: string
  // Required by the backend whenever item_type is 'raw_material' and
  // movement_type is 'receipt' -- see backend/app/schemas/inventory.py.
  supplier_id?: number | null
  unit_cost?: number | null
  batch_number?: string | null
  expiry_date?: string | null
  invoice_number?: string | null
  received_by?: string | null
  received_date?: string | null
}

/** 'applied' | 'pending_approval' -- see
 * backend/app/services/inventory_service.py's submit_manual_adjustment. */
export interface StockAdjustmentResult {
  status: 'applied' | 'pending_approval'
  stock: StockLevel | null
  request: StockAdjustmentRequest | null
}

export type StockAdjustmentRequestStatus = 'pending' | 'applied' | 'rejected'

export interface StockAdjustmentRequest {
  id: number
  item_type: InventoryItemType
  item_id: number
  quantity: number
  movement_type: MovementType
  reason: string
  supplier_id: number | null
  unit_cost: number | null
  batch_number: string | null
  expiry_date: string | null
  invoice_number: string | null
  received_by: string | null
  received_date: string | null
  status: StockAdjustmentRequestStatus
  rejection_reason: string | null
  resulting_movement_id: number | null
  requested_by: number | null
  requested_at: string
  decided_by: number | null
  decided_at: string | null
}

export interface LowStockItem {
  raw_material_id: number
  code: string
  name: string
  quantity_on_hand: number
  reorder_point: number
}

export interface FinishedGoodStockItem {
  product_id: number
  code: string
  name: string
  unit: string
  product_status: 'active' | 'inactive'
  quantity_on_hand: number
  quantity_reserved: number
  quantity_available: number
  reorder_point: number
  is_low: boolean
}

/** 'raw_material' | 'packaging' | 'consumable' -- see
 * backend/app/models/raw_material.py's material_type. */
export type RawMaterialType = 'raw_material' | 'packaging' | 'consumable'

export interface RawMaterialStockItem {
  raw_material_id: number
  code: string
  name: string
  unit: string
  material_type: RawMaterialType
  material_status: 'active' | 'inactive'
  quantity_on_hand: number
  quantity_reserved: number
  quantity_available: number
  reorder_point: number
  is_low: boolean
  // Outstanding quantity on an open purchase order -- always a real
  // number (0 when nothing's on order).
  incoming_quantity: number
  // From MRP -- only present when this material currently has live
  // production demand and a shortfall against on-hand stock. Null means
  // "no known requirement right now", not "confirmed zero demand".
  required_quantity: number | null
  shortfall: number | null
}

export interface StockMovement {
  id: number
  item_type: InventoryItemType
  item_id: number
  item_name: string | null
  item_route: string | null
  movement_type: AnyMovementType
  quantity: number
  reference_type: string | null
  reference_id: number | null
  reference_label: string | null
  reference_route: string | null
  supplier_id: number | null
  unit_cost: number | null
  batch_number: string | null
  expiry_date: string | null
  invoice_number: string | null
  received_by: string | null
  received_date: string | null
  notes: string | null
  created_at: string
  created_by: number | null
}

File: backend/app/schemas/inventory.py
from datetime import date, datetime

from pydantic import BaseModel, Field


class StockLevelOut(BaseModel):
    item_type: str
    item_id: int
    quantity_on_hand: float
    quantity_reserved: float
    quantity_available: float
    # Set only when this StockLevelOut was returned by a call that just
    # created a movement (adjust_stock) -- the ledger entry this exact
    # adjustment is traceable to. None from a plain GET /stock/{type}/{id}.
    movement_id: int | None = None


class StockAdjustRequest(BaseModel):
    item_type: str = Field(pattern="^(product|raw_material)$")
    item_id: int
    quantity: float = Field(description="Positive = stock in, negative = stock out")
    movement_type: str = Field(pattern="^(receipt|issue|adjustment|return)$")
    # Mandatory for every manual adjustment -- see
    # inventory_service.submit_manual_adjustment.
    notes: str = Field(min_length=1, max_length=255)
    # Required by inventory_service.adjust_stock whenever item_type is
    # 'raw_material' and movement_type is 'receipt' -- a raw material
    # arriving at the factory with no supplier, cost, invoice, receiver,
    # or date attached is exactly the missing data that breaks supplier
    # and cost analytics later. batch_number/expiry_date stay optional --
    # not every raw material is batch or expiry tracked.
    supplier_id: int | None = None
    unit_cost: float | None = Field(default=None, ge=0)
    batch_number: str | None = None
    expiry_date: date | None = None
    invoice_number: str | None = None
    received_by: str | None = None
    received_date: date | None = None


class LowStockItem(BaseModel):
    raw_material_id: int
    code: str
    name: str
    quantity_on_hand: float
    reorder_point: float


class FinishedGoodStockItem(BaseModel):
    product_id: int
    code: str
    name: str
    unit: str
    product_status: str
    quantity_on_hand: float
    quantity_reserved: float
    quantity_available: float
    reorder_point: float
    is_low: bool

    model_config = {"from_attributes": True}


class RawMaterialStockItem(BaseModel):
    raw_material_id: int
    code: str
    name: str
    unit: str
    # 'raw_material' | 'packaging' | 'consumable' -- see
    # RawMaterial.material_type. Lets the raw-materials stock view be
    # filtered to just packaging (or just consumables), shown
    # independently from ordinary raw material stock.
    material_type: str
    material_status: str
    quantity_on_hand: float
    quantity_reserved: float
    quantity_available: float
    reorder_point: float
    is_low: bool

    model_config = {"from_attributes": True}


class StockMovementOut(BaseModel):
    id: int
    item_type: str
    item_id: int
    movement_type: str
    quantity: float
    reference_type: str | None
    reference_id: int | None
    supplier_id: int | None
    unit_cost: float | None
    batch_number: str | None
    expiry_date: date | None
    invoice_number: str | None
    received_by: str | None
    received_date: date | None
    notes: str | None
    created_at: datetime
    created_by: int | None

    model_config = {"from_attributes": True}


class StockAdjustmentRequestOut(BaseModel):
    id: int
    item_type: str
    item_id: int
    quantity: float
    movement_type: str
    reason: str
    supplier_id: int | None
    unit_cost: float | None
    batch_number: str | None
    expiry_date: date | None
    invoice_number: str | None
    received_by: str | None
    received_date: date | None
    status: str
    rejection_reason: str | None
    resulting_movement_id: int | None
    requested_by: int | None
    requested_at: datetime
    decided_by: int | None
    decided_at: datetime | None

    model_config = {"from_attributes": True}


class StockAdjustmentResultOut(BaseModel):
    # 'applied' | 'pending_approval' -- see
    # inventory_service.submit_manual_adjustment.
    status: str
    stock: StockLevelOut | None
    request: StockAdjustmentRequestOut | None


class StockAdjustmentRejectRequest(BaseModel):
    reason: str = Field(min_length=1, max_length=5000)
